Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs

```diff
@@ -125,82 +124,54 @@ public bool DisconnectDevice(EntityUid uid, DeviceNetworkComponent? device, bool
-                return devices;
-
-            return new();
-        }
-
-        /// <summary>
-        /// Get a list of devices listening for ANY transmission on a given frequency, rather than just broadcast & addressed events.
-        /// </summary>
-        private HashSet<DeviceNetworkComponent> GetRecieveAllDevices(ConnectionType netId, uint frequency)
-        {
-            if (_networks.TryGetValue(netId, out var network) && network.ReceiveAllDevices.TryGetValue(frequency, out var devices))
-                return devices;
-
-            return new();
+            return GetNetwork(device.DeviceNetId).Remove(device);
         }
 
         /// <summary>
         ///     Try to find a device on a network using its address.
         /// </summary>
-        private bool TryGetDevice(ConnectionType netId, string address, [NotNullWhen(true)] out DeviceNetworkComponent? device)
-        {
-            if (!_networks.TryGetValue(netId, out var network))
-            {
-                device = null;
-                return false;
-            }
-
-            return network.Devices.TryGetValue(address, out device);
-        }
-        #endregion
+        private bool TryGetDevice(ConnectionType netId, string address, [NotNullWhen(true)] out DeviceNetworkComponent? device) =>
+            GetNetwork(netId).Devices.TryGetValue(address, out device);
 
-        #region Packet Sending
         private void SendPacket(DeviceNetworkPacketEvent packet)
         {
-            HashSet<DeviceNetworkComponent> recipients;
-
+            var network = GetNetwork(packet.NetId);
             if (packet.Address == null)
             {
-                // Broadcast to all listening devices
-                recipients = GetListeningDevices(packet.NetId, packet.Frequency);
+                if (network.ListeningDevices.TryGetValue(packet.Frequency, out var devices))
+                {
+                    var deviceCopy = ArrayPool<DeviceNetworkComponent>.Shared.Rent(devices.Count);
+                    devices.CopyTo(deviceCopy);
+                    SendToConnections(deviceCopy, packet);
+                    ArrayPool<DeviceNetworkComponent>.Shared.Return(deviceCopy);
+                }
             }
             else
             {
-                // Add devices listening to all messages
-                recipients = new(GetRecieveAllDevices(packet.NetId, packet.Frequency));
-
-                // add the intended recipient (if they are even listening).
+                var totalDevices = 0;
+                if (network.ReceiveAllDevices.TryGetValue(packet.Frequency, out var devices))
+                {
+                    totalDevices += devices.Count;
+                }
                 if (TryGetDevice(packet.NetId, packet.Address, out var device) && device.ReceiveFrequency == packet.Frequency)
-                    recipients.Add(device);
+                {
+                    totalDevices += 1;
+                }
```

Comment:
now that its not a hashset you have to check whether the intended recipient has receive all set to true, because otherwise it gets included twice

Reply:
It is checked now